src/transform/src/normalize_lets.rs

```diff
@@ -0,0 +1,360 @@
+        let renumber = to_emit
+            .keys()
+            .enumerate()
+            .any(|(i, b)| LocalId::new(i as u64) != *b);
+
+        // Reconstitute the stack of let bindings from `to_emit`.
+        for (id, value) in to_emit.into_iter().rev() {
+            *relation = MirRelationExpr::Let {
+                id,
+                value: Box::new(value),
+                body: Box::new(relation.take_dangerous()),
+            };
+        }
+
+        // Do a final pass if we inlined anything, in order to count from zero.
+        if renumber {
+            let mut id_gen = IdGen::default();
+            renumber_bindings(relation, &mut id_gen)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// Re-assign an identifier to each `Let`.
+pub fn renumber_bindings(
+    relation: &mut MirRelationExpr,
+    id_gen: &mut IdGen,
+) -> Result<(), crate::TransformError> {
+    let mut renaming = BTreeMap::new();
+    let recursion_guard = RecursionGuard::with_limit(RECURSION_LIMIT);
+    support::renumber_bindings_helper(&recursion_guard, relation, &mut renaming, id_gen)
+}
+
+// Support methods that are unlikely to be useful to other modules.
+mod support {
+
+    use std::collections::BTreeMap;
+
+    use mz_expr::{Id, LocalId, MirRelationExpr};
+    use mz_ore::id_gen::IdGen;
+    use mz_ore::stack::{CheckedRecursion, RecursionGuard};
+
+    // This is pretty simple, but it is used in two locations and seemed clearer named.
+    /// Populates `counts` with the number of uses of each local identifier.
+    pub(super) fn count_local_id_uses(
+        expr: &MirRelationExpr,
+        counts: &mut std::collections::BTreeMap<LocalId, usize>,
+    ) {
+        expr.visit_pre(|expr| {
+            if let MirRelationExpr::Get {
+                id: Id::Local(i), ..
+            } = expr
+            {
+                *counts.entry(*i).or_insert(0) += 1;
+            }
+        });
+    }
+
+    /// Extract all `Let` bindings from an expression, into a map from `LocalId` to `Let`-free expressions.
+    ///
+    /// An equivalent expression can be reconstructed by introducing the bindings atop the resulting `expr`.
+    pub(super) fn digest_lets(expr: &mut MirRelationExpr) -> BTreeMap<LocalId, MirRelationExpr> {
+        let mut lets = BTreeMap::new();
+        let mut worklist = Vec::new();
+        digest_lets_helper(expr, &mut worklist);
+        while let Some((id, mut expr)) = worklist.pop() {
+            digest_lets_helper(&mut expr, &mut worklist);
+            lets.insert(id, expr);
+        }
+        lets
+    }
+
+    /// Extract all `Let` bindings from `expr`, into `(id, value)` pairs.
+    ///
+    /// Importantly, the `value` pairs may not be `Let`-free, and must be further processed.
+    fn digest_lets_helper(
+        expr: &mut MirRelationExpr,
+        bindings: &mut Vec<(LocalId, MirRelationExpr)>,
+    ) {
+        let mut to_visit = vec![expr];
+        while let Some(expr) = to_visit.pop() {
```

Comment:
The `Let` case re-inserts `expr` into the stack rather than directly continuing recursively with its children. I think there was a visitor variant that allowed you to specify how to continue the visitation, but I find the five lines here are clearer than the closures required to direct that pattern.